tinygraph/tinygraph-tests.c

```diff
@@ -1124,6 +1124,239 @@ void test33(void) {
   assert(tinygraph_is_sorted_u32(a2, sizeof(a2) / sizeof(a2[0])));
 }
 
+void test34(void) {
+  tinygraph_rng_s rng = tinygraph_rng_construct();
+
+  const uint32_t n = 1000;
+
+  uint32_t * const nodes = malloc(n * sizeof(uint32_t));
+  uint16_t * const lngs = malloc(n * sizeof(uint16_t));
+  uint16_t * const lats = malloc(n * sizeof(uint16_t));
+
+  assert(nodes);
+  assert(lngs);
+  assert(lats);
+
+  tinygraph_iota_u32(nodes, n, 0);
+
+  for (uint32_t i = 0; i < n; ++i) {
+    lngs[i] = tinygraph_rng_random(rng);
+    lats[i] = tinygraph_rng_random(rng);
+  }
+
+  const bool ok = tinygraph_reorder(nodes, lngs, lats, n);
+  assert(ok);
+
+  free(nodes);
+  free(lngs);
+  free(lats);
+
+  tinygraph_rng_destruct(rng);
+}
+
+
+TINYGRAPH_WARN_UNUSED
+tinygraph_s construct_random_graph(tinygraph_rng_s rng, uint32_t nodes, uint32_t degree) {
+  assert(rng);
+  assert(nodes > 0);
+  assert(degree > 1);
+
+  const uint32_t edges = nodes * degree;
+
+  uint32_t * const sources = malloc(edges * sizeof(uint32_t));
+
+  if (!sources) {
+    return NULL;
+  }
+
+  uint32_t * const targets = malloc(edges * sizeof(uint32_t));
+
+  if (!targets) {
+    free(sources);
+    return NULL;
+  }
+
+  for (uint32_t i = 0; i < nodes; ++i) {
+    for (uint32_t j = 0; j < degree; ++j) {
+      const uint32_t source = i;
+
+      const uint32_t target = tinygraph_rng_bounded(rng, nodes);
+
+      sources[i * degree + j] = source;
+      targets[i * degree + j] = target;
+    }
+  }
+
+  tinygraph_s graph = tinygraph_construct_from_unsorted_edges(
+      sources, targets, edges);
+
+  free(sources);
+  free(targets);
+
+  return graph;
+}
+
+TINYGRAPH_WARN_UNUSED
+tinygraph_s construct_embedded_graph(tinygraph_rng_s rng, uint32_t nodes, uint32_t degree) {
+  assert(rng);
+  assert(nodes > 0);
+  assert(degree > 1);
+
+  const uint32_t edges = nodes * degree;
+
+  uint32_t * const sources = malloc(edges * sizeof(uint32_t));
+
+  if (!sources) {
+    return NULL;
+  }
+
+  uint32_t * const targets = malloc(edges * sizeof(uint32_t));
+
+  if (!targets) {
+    free(sources);
+    return NULL;
+  }
+
+  // Simulate what we will get with a reordering based on a
+  // spatial embedding: edges (s, t) where: |s-t| is min.
+
+  for (uint32_t i = 0; i < nodes; ++i) {
+    for (uint32_t j = 0; j < degree; ++j) {
+      const uint32_t source = i;
+
+      uint32_t f = tinygraph_rng_bounded(rng, degree * 2);
+      uint32_t b = tinygraph_rng_bounded(rng, degree * 2);
+
+      uint32_t target = i - f + b;
+      target = target >= nodes ? i : target;
```

Comment:
Probably not the best heuristic to generate somewhat local edges but good enough for now.